src/util/specific.py

```diff
@@ -137,3 +138,66 @@ def reset_cowed_views(cr, xmlid, key=None):
         [key, module, name],
     )
     return set(sum(cr.fetchall(), ()))
+
+
+def translation2jsonb_unofficial_fields(cr):
```

Comment:
I'd suggest you restrict the fields to be upgraded. Or maybe allow also to restrict by models. It would avoid unintentional changes and be faster if you can callers pinpoint the fields that need a fix (maybe from the info in the upgrade report). 

Also assert that the running DB is at least in 16.0.

Something like:
```suggestion
def convert_custom_translated_fields(cr, models=None, fields=None):
    assert util.version_gte("16.0")
    assert bool(models) ^ bool(fields)
    if models:
        # fix all fields per model that look like they weren't yet updated
    else:
        # fix specific fields = [(model_name, field_name), ...]
```
You could identify if a field looks like it wasn't updated by inspecting its column. If the jsonb dict only has en_US and lacks some langs for which there are translations in _ir_translation then it's most probably something custom. Note that I still believe it's really better to restrict by specific fields instead of per model. I also understand that for huge custom models (with many fields) this may be cumbersome, hence the alternative to restrict by model.

FYI on custom fields we _only_ run `util.convert_field_to_translatable(cr, model_name, field)`.

Reply:
> Maintaining a hardcoded list of modules, models or fields to translate? No, thanks. Everything is already in the database. At least, everything we care about (user-generated translations).

If you can be sure you can rely on the info the DB to pick the data then this is fine. In our experience relying only on the DB is error prone, and fails for many (common) edge cases (like wrong data in the DB). All out internal upgrade scripts are specific about modules or fields updates. An upgrade script is something you write once for your module and should be stable between major versions. I don't see any problem in having a full list of models and fields you need to update. It would be exceptional to have hundreds of **translated** fields in a module, and even in such case having a script that you write once is not a problem.

I understand that what you are trying to add here is a "general" function to update all fields. My point is that it's better to add the upgrade script to each module in the right folder. It's each module's responsibility to update their data. In short, I think `translation2jsonb` is fine, `translation2jsonb_all_missing` is a bit problematic.

Note: I didn't check the code yet, this is a reply to your comment only.

Reply:
It depends on your use case.

You guys massively migrate thousands of DBs per day, with no control over the input but full control over the migration scope (Odoo CE+EE+design-themes+industry addons; DB schema frozen at launch day). I understand it makes sense for you indeed.

We are more in the retail side. We do have control over the input, so if a translation is wrong there, we can just change it there and migrate again. But we have limited control over migration scope (a new customer can hire us, with an old Odoo installation with a big amount of custom modules; some can be communitary and still get updates 10 years later).

So, I could remove `translation2jsonb_all_missing` from the PR, but at the end most out-of-Odoo people that will use upgrade-util and `translation2jsonb` will write something similar on their own. You choose, anyway.

FWIW, [OpenUpgrade does maintain a list of excluded fields](https://github.com/OCA/OpenUpgrade/blob/2e9d96f1288596f17d5af3fad84b86eca6dc3821/openupgrade_scripts/scripts/base/16.0.1.3/end-migration.py#L9-L39). That's IMHO better suited for downstream scripts.

Reply:
This is a nice perspective. We are all discovering the best approach for this repo. Thanks for the details. For us it's a bit hard to foresee the community usage and so we will try to gain insights like this one.

For the moment I think this patch looks good. I may have some nitpick comments (still review in progress) but it looks good enough.